### DISCUSSION TITLE
Optimizing Splink Performance on a High-Resource Single Machine - Tips and Tricks

### DISCUSSION THREAD

**Discussion by alexwoolford** [moj-analytical-services/splink#1150] [Q&A] [answered]
I tried to scale-out Splink with Spark and Athena but faced difficulties with deployment and debugging. As an alternative, I decided to try a simpler approach with a single powerful machine.

I spun up an r6a.16xlarge on AWS (64 cores, 512GB instance), installed the prerequisite libraries (Splink, Jupyter, drivers to connect to Snowflake, etc.), estimated the u/m probabilities, and then called `linker.predict(threshold_match_probability=0.9)`.

Given the instance specs, I had hoped that a VM of this size would be several times faster than my laptop. However, it didn't yield the step change in performance that I'd hoped for. Using `htop`, I found that while there are 64 cores on the box, only 11 of them were used in the call to `linker.estimate_m_from_label_column("mycol")`, and just 3 cores were pegged whilst running linker.predict(...).

Questions:
1. Is "Splink on a big box" an anti-pattern?
2. Are there any tricks/tweaks we can employ to take advantage of the hardware resources available?
3. Is there a way to configure Splink to use more cores or parallelize the computations better to improve performance on a large instance like this?

Any insights or recommendations would be greatly appreciated.

**Comment by RobinL** [comment, marked as answer]
Interesting!  Is that DuckDB?  We haven't done too much testing on DuckDB but I have noticed similar.

On `predict()`, how many blocking rules are in your list of `blocking_rules_for_prediction`?  My instinct would suggest that the more there are, the better parallelisation you'll get (that's certainly true on the Spark side).  But I haven't tested on DuckDB.

We could also consider activating salting for DuckDB (which is a suggested solution for these issues in Spark) https://moj-analytical-services.github.io/splink/topic_guides/salting.html

**Reply by alexwoolford** [reply]
Thanks, Robin. It's an honor.

It was DuckDB. And I have 9 blocking rules.

Your comment about salting inspired me to revisit running Splink on Spark. It worked perfectly on my single, beefy instance. I matched the number of salting partitions to the number of cores, and then watched the instance spring into life:

```
"blocking_rules_to_generate_predictions": [
    {"blocking_rule": "l.COMPANY_NAME = r.COMPANY_NAME", "salting_partitions": 64},
],
```

I had previously struggled with Splink on EMR (particularly displaying the Vega-Lite plots in Sagemaker's flavor of Jupyter). While scale-up feels like an anti-pattern, it got the job done. 😄

**Reply by RobinL** [reply]
Great - also thanks very much I appreciate you taking the time to update us
